ncclient/transport/ssh.py

```diff
@@ -149,149 +163,84 @@ def _parse10(self):
-                else:
-                    inendpos += 1 # == 3 now #
-                    if x == b'#':
-                        state = inend
-                    elif x.isdigit():
-                        # More trunks
-                        state = instart
-                        num_list = []
-                        num_list.append(x)
-                    else:
-                        logger.debug('%s (%s: expect "#")'%(pre, state))
-                        raise Exception
-            elif state == inend:
-                if inendpos == 3:
-                    if x == b'\n':
-                        inendpos = 0
-                        state = idle
-                        logger.debug('dispatching message')
-                        self._dispatch_message(''.join(message_list))
-                        # reset
-                        rest = buf.read()
-                        buf = BytesIO()
-                        buf.write(rest)
-                        buf.seek(0)
-                        message_list = []
-                        self._message_list = message_list
-                        chunk_list = []
-                        expchunksize = chunksize = 0
-                        parsing_state11 = idle
-                        inendpos = parsing_pos11 = 0
-                        # There could be another entire message in the
-                        # buffer, so we should try to parse again.
-                        should_recurse = True
-                        break
-                    else:
-                        logger.debug('%s (%s: expect newline)'%(pre, state))
-                        raise Exception
-            else:
-                logger.debug('%s (%s invalid state)'%(pre, state))
-                raise Exception
-
-        self._expchunksize = expchunksize
-        self._curchunksize = curchunksize
-        self._parsing_state11 = state
-        self._inendpos = inendpos
-        self._size_num_list = num_list
-        self._buffer = buf
-        self._parsing_pos11 = self._buffer.tell()
-        logger.debug('parse11 ending ...')
-
-        if should_recurse:
-            logger.debug('Trying another round of parsing since there is still data')
+                    # we don't have enough bytes, just break out for now
+                    # after updating start pointer to start of new chunk
+                    start += re_start
+                    logger.debug('_parse11: not enough data for chunk yet')
+                    logger.debug('_parse11: setting start to {0}'.format(start))
+                    break
+                
+        # Now out of the loop, need to see if we need to save back any content
+        if start > 0:
+            logger.debug('_parse11: saving back rest of message after {0} bytes, original size {1}'.format(start, data_len))
+            self._buffer = StringIO(data[start:])
+            logger.debug('_parse11: still have data, may have another full message!')
```

Comment:
Possibly add a check here on either `len(self._buffer) > 0` or `start < data_len` before recursing? Here's what I see in the logs, which is a bit suboptimal:

```
_parse11: saving back rest of message after 2640 bytes, original size 2640
_parse11: still have data, may have another full message!
_parse11: starting
_parse11: working with buffer of 0 bytes
_parse11: matching from 0 bytes from start of buffer
_parse11: no delimiter found, buffer=b''
_parse11: ending
```

Reply:
Yep, now looks like:

```
        if start > 0:
            logger.debug('_parse11: saving back rest of message after {0} bytes, original size {1}'.format(start, data_len))
            self._buffer = StringIO(data[start:])
            if start < data_len:
                logger.debug('_parse11: still have data, may have another full message!')
                self._parse11()
```